File: packages/core/src/policy/policy-loader.ts
/**
 * Policy-as-YAML loader for Matimo.
 *
 * Allows the developer to configure the policy engine through a YAML file
 * instead of inline TypeScript, making it easy to adjust policy across
 * environments without rebuilding.
 *
 * Schema for policy.yaml:
 *
 * ```yaml
 * allowedDomains:
 *   - api.slack.com
 *   - slack.com
 *
 * allowedCredentials:
 *   - SLACK_BOT_TOKEN
 *   - OPENAI_API_KEY
 *
 * allowedHttpMethods:
 *   - GET
 *   - POST
 *
 * allowCommandTools: false
 * allowFunctionTools: false
 *
 * protectedNamespaces:
 *   - matimo_
 * ```
 *
 * Usage:
 *   const matimo = await MatimoInstance.init({ policyFile: './policy.yaml' });
 */

import fs from 'fs';
import yaml from 'js-yaml';
import { z } from 'zod';
import { DefaultPolicyEngine } from './default-policy';
import type { PolicyEngine, PolicyConfig } from './types';
import { MatimoError, ErrorCode } from '../errors/matimo-error';

// ──────────────────────────────────────────────────────────────────────────────
// Zod schema — validates the YAML before constructing PolicyConfig
// ──────────────────────────────────────────────────────────────────────────────

// Valid HTTP methods supported by the policy engine
const VALID_HTTP_METHODS = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'HEAD', 'OPTIONS'] as const;
const ValidHttpMethodEnum = z.enum(VALID_HTTP_METHODS);

const PolicyFileSchema = z.object({
  allowedDomains: z.array(z.string()).optional(),
  allowedCredentials: z.array(z.string()).optional(),
  allowedHttpMethods: z
    .array(
      z
        .string()
        .transform((val: string) => val.toUpperCase())
        .pipe(ValidHttpMethodEnum)
    )
    .optional(),
  allowCommandTools: z.boolean().optional(),
  allowFunctionTools: z.boolean().optional(),
  protectedNamespaces: z.array(z.string()).optional(),
  enableHITL: z.boolean().optional(),
  quarantineRiskLevels: z.array(z.enum(['low', 'medium', 'high', 'critical'])).optional(),
});

type PolicyFile = z.infer<typeof PolicyFileSchema>;

// ──────────────────────────────────────────────────────────────────────────────
// Public API
// ──────────────────────────────────────────────────────────────────────────────

/**
 * Parse a YAML policy file and return a PolicyEngine configured from it.
 *
 * Throws `MatimoError(INVALID_SCHEMA)` if the file cannot be read or fails validation.
 *
 * @param filePath - Absolute or cwd-relative path to the policy YAML file
 * @returns A frozen `DefaultPolicyEngine` built from the parsed config
 *
 * @example
 * ```ts
 * // Direct usage
 * const engine = loadPolicyFromFile('./policy.yaml');
 * const matimo = await MatimoInstance.init({ policy: engine });
 *
 * // Or use the shorthand InitOption (preferred)
 * const matimo = await MatimoInstance.init({ policyFile: './policy.yaml' });
 * ```
 */
export function loadPolicyFromFile(filePath: string): PolicyEngine {
  let raw: string;
  try {
    raw = fs.readFileSync(filePath, 'utf-8');
  } catch (err) {
    throw new MatimoError(
      `Cannot read policy file "${filePath}": ${(err as NodeJS.ErrnoException).message}`,
      ErrorCode.INVALID_SCHEMA,
      { filePath }
    );
  }

  let parsed: unknown;
  try {
    parsed = yaml.load(raw);
  } catch (err) {
    throw new MatimoError(
      `Policy file "${filePath}" contains invalid YAML: ${(err as Error).message}`,
      ErrorCode.INVALID_SCHEMA,
      { filePath }
    );
  }

  const result = PolicyFileSchema.safeParse(parsed ?? {});
  if (!result.success) {
    const issues = result.error.issues
      .map((i) => `  • ${i.path.join('.')}: ${i.message}`)
      .join('\n');
    throw new MatimoError(
      `Policy file "${filePath}" is invalid:\n${issues}`,
      ErrorCode.INVALID_SCHEMA,
      { filePath, issues: result.error.issues }
    );
  }

  const policyConfig: PolicyConfig = buildPolicyConfig(result.data);
  return new DefaultPolicyEngine(policyConfig);
}

/**
 * Parse a YAML policy file into a PolicyConfig (without creating an engine).
 * Useful for hot-reload: parse the new file, then call engine.updateConfig().
 */
export function parsePolicyFile(filePath: string): PolicyConfig {
  let raw: string;
  try {
    raw = fs.readFileSync(filePath, 'utf-8');
  } catch (err) {
    throw new MatimoError(
      `Cannot read policy file "${filePath}": ${(err as NodeJS.ErrnoException).message}`,
      ErrorCode.INVALID_SCHEMA,
      { filePath }
    );
  }

  let parsed: unknown;
  try {
    parsed = yaml.load(raw);
  } catch (err) {
    throw new MatimoError(
      `Policy file "${filePath}" contains invalid YAML: ${(err as Error).message}`,
      ErrorCode.INVALID_SCHEMA,
      { filePath }
    );
  }

  const result = PolicyFileSchema.safeParse(parsed ?? {});
  if (!result.success) {
    const issues = result.error.issues
      .map((i) => `  \u2022 ${i.path.join('.')}: ${i.message}`)
      .join('\n');
    throw new MatimoError(
      `Policy file "${filePath}" is invalid:\n${issues}`,
      ErrorCode.INVALID_SCHEMA,
      { filePath, issues: result.error.issues }
    );
  }

  return buildPolicyConfig(result.data);
}

function buildPolicyConfig(data: PolicyFile): PolicyConfig {
  const config: PolicyConfig = {};
  if (data.allowedDomains !== undefined) config.allowedDomains = data.allowedDomains;
  if (data.allowedCredentials !== undefined) config.allowedCredentials = data.allowedCredentials;
  if (data.allowedHttpMethods !== undefined) config.allowedHttpMethods = data.allowedHttpMethods;
  if (data.allowCommandTools !== undefined) config.allowCommandTools = data.allowCommandTools;
  if (data.allowFunctionTools !== undefined) config.allowFunctionTools = data.allowFunctionTools;
  if (data.protectedNamespaces !== undefined) config.protectedNamespaces = data.protectedNamespaces;
  if (data.enableHITL !== undefined) config.enableHITL = data.enableHITL;
  if (data.quarantineRiskLevels !== undefined)
    config.quarantineRiskLevels = data.quarantineRiskLevels;
  return config;
}

File: python/packages/core/src/matimo/policy/policy_loader.py
"""
Policy loader — parses policy.yaml and returns a DefaultPolicyEngine.
Mirrors: packages/core/src/policy/policy-loader.ts
"""
from __future__ import annotations

import logging
from pathlib import Path

import yaml
from pydantic import ValidationError

from matimo.errors import ErrorCode, MatimoError
from matimo.policy.default_policy import DefaultPolicyEngine
from matimo.policy.types import PolicyConfig, RiskLevel

logger = logging.getLogger("matimo")


def load_policy_from_file(
    file_path: str | Path,
    trusted_paths: list[str] | None = None,
    untrusted_paths: list[str] | None = None,
) -> DefaultPolicyEngine:
    """
    Load a policy.yaml file and return a configured DefaultPolicyEngine.

    Expected YAML keys (all optional — safe defaults if omitted):
      allowedDomains: [...]
      allowedCredentials: [...]
      allowedHttpMethods: [GET, POST]
      allowCommandTools: false
      allowFunctionTools: false
      protectedNamespaces: [matimo_]
      enableHITL: false
      quarantineRiskLevels: [medium]

    Raises:
        MatimoError(FILE_NOT_FOUND) if the file does not exist.
        MatimoError(INVALID_SCHEMA) if the YAML is malformed.
    """
    path = Path(file_path)
    if not path.exists():
        raise MatimoError(
            f"Policy file not found: {path}",
            ErrorCode.FILE_NOT_FOUND,
            {"path": str(path)},
        )

    try:
        raw = path.read_text(encoding="utf-8")
        data = yaml.safe_load(raw) or {}
    except Exception as exc:
        raise MatimoError(
            f"Failed to parse policy file: {path}\n{exc}",
            ErrorCode.INVALID_SCHEMA,
            {"path": str(path)},
            cause=exc,
        ) from exc

    # Normalise camelCase YAML keys → snake_case for Pydantic
    normalised = _normalise_keys(data)

    try:
        config = PolicyConfig.model_validate(normalised)
    except ValidationError as exc:
        raise MatimoError(
            f"Policy config validation failed: {path}\n{exc}",
            ErrorCode.INVALID_SCHEMA,
            {"path": str(path), "issues": exc.errors()},
            cause=exc,
        ) from exc

    logger.info("Loaded policy from %s", path)
    return DefaultPolicyEngine(
        config=config,
        trusted_paths=trusted_paths,
        untrusted_paths=untrusted_paths,
    )


# ---------------------------------------------------------------------------
# Key normalisation helper
# ---------------------------------------------------------------------------

_CAMEL_TO_SNAKE: dict[str, str] = {
    "allowedDomains": "allowed_domains",
    "allowedCredentials": "allowed_credentials",
    "allowedHttpMethods": "allowed_http_methods",
    "allowCommandTools": "allow_command_tools",
    "allowFunctionTools": "allow_function_tools",
    "protectedNamespaces": "protected_namespaces",
    "enableHITL": "enable_hitl",
    "quarantineRiskLevels": "quarantine_risk_levels",
    "approvalTtlSeconds": "approval_ttl_seconds",
}


def _normalise_keys(data: dict) -> dict:
    """Convert camelCase policy.yaml keys to snake_case for Pydantic."""
    result: dict = {}
    for k, v in data.items():
        snake_key = _CAMEL_TO_SNAKE.get(k, k)
        if snake_key == "quarantine_risk_levels" and isinstance(v, list):
            v = [RiskLevel(level) for level in v]
        result[snake_key] = v
    return result
